File: bank-service/src/main/java/rs/edu/raf/service/impl/ExchangeInvoiceServiceImpl.java
package rs.edu.raf.service.impl;

import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;
import rs.edu.raf.annotations.GeneratedCrudOperation;
import rs.edu.raf.dto.ExchangeInvoiceDTO;
import rs.edu.raf.mapper.ExchangeInvoiceMapper;
import rs.edu.raf.model.ExchangeInvoice;
import rs.edu.raf.model.accounts.BankAccount;
import rs.edu.raf.repository.ExchangeInvoiceRepository;
import rs.edu.raf.service.ExchangeInvoiceService;
import rs.edu.raf.service.ExchangeRateService;

import java.math.BigDecimal;
import java.util.List;

@Service
@AllArgsConstructor
public class ExchangeInvoiceServiceImpl implements ExchangeInvoiceService {
    private ExchangeInvoiceRepository exchangeInvoiceRepository;
    private ExchangeInvoiceMapper exchangeInvoiceMapper;
    private ExchangeRateService exchangeRateService;

    @Override
    public void createInvoiceForExchange(BankAccount sender, BankAccount receiver, BigDecimal amount) {
        ExchangeInvoice exchangeInvoice = new ExchangeInvoice();
        exchangeInvoice.setSenderAccount(sender);
        exchangeInvoice.setReceiverAccount(receiver);
        exchangeInvoice.setSenderAmount(amount);
        exchangeInvoice.setSenderCurrency(sender.getCurrency());
        exchangeInvoice.setReceiverCurrency(receiver.getCurrency());
        exchangeInvoice.setExchangeRate(exchangeRateService.exchangeRate(sender.getCurrency().getCode(), receiver.getCurrency().getCode()));
        exchangeInvoice.setProfit(amount.multiply(new BigDecimal("0.005")));
        exchangeInvoice.setDateAndTime(System.currentTimeMillis());
        exchangeInvoiceRepository.save(exchangeInvoice);
    }

    @GeneratedCrudOperation
    @Override
    public List<ExchangeInvoiceDTO> listInvoicesByCurrency(String currency) {
        return exchangeInvoiceRepository.findAllBySenderCurrency_Code(currency)
                .stream()
                .map(exchangeInvoiceMapper::exchangeInvoiceToExchangeInvoiceDTO)
                .toList();
    }
}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/service/impl/ExchangeInvoiceServiceImpl.kt
package org.example.bankservicekotlin.service.impl

import org.example.bankservicekotlin.dto.ExchangeInvoiceDTO
import org.example.bankservicekotlin.mapper.ExchangeInvoiceMapper
import org.example.bankservicekotlin.model.ExchangeInvoice
import org.example.bankservicekotlin.model.accounts.BankAccount
import org.example.bankservicekotlin.repository.ExchangeInvoiceRepository
import org.example.bankservicekotlin.service.ExchangeInvoiceService
import org.example.bankservicekotlin.service.ExchangeRateService
import org.springframework.stereotype.Service
import java.math.BigDecimal

@Service
class ExchangeInvoiceServiceImpl(
    private val exchangeInvoiceRepository: ExchangeInvoiceRepository,
    private val exchangeInvoiceMapper: ExchangeInvoiceMapper,
    private val exchangeRateService: ExchangeRateService
) : ExchangeInvoiceService {
    override fun createInvoiceForExchange(sender: BankAccount, receiver: BankAccount, amount: BigDecimal) {
        val exchangeRate = exchangeRateService.exchangeRate(
            sender.currency.code,
            receiver.currency.code
        )
        val exchangeInvoice = ExchangeInvoice(
            null,
            sender,
            receiver,
            amount,
            sender.currency,
            receiver.currency,
            exchangeRate,
            amount.multiply(BigDecimal("0.005")),
            System.currentTimeMillis()
        )
        exchangeInvoiceRepository.save(exchangeInvoice)
    }

    override fun listInvoicesByCurrency(currency: String): List<ExchangeInvoiceDTO> {
        return exchangeInvoiceRepository.findAllBySenderCurrency_Code(currency)
            .stream()
            .map(exchangeInvoiceMapper::exchangeInvoiceToExchangeInvoiceDTO)
            .toList()
    }
}